File: QuesFromTest/01/n-numbers-with-max-product.cpp
/* Author: Jitendra Kumar (jitendra.iitp@gmail.com) */
#include<bits/stdc++.h>
using namespace std;
#define endl '\n'
#define MOD 1000000007
#define DEBUG(x) cerr << #x << " is " << (x) << endl;

void pick_N_numbers(vector<int> arr, int numberOfElements, int N, vector<int> &pickedNumbers) {
    if(N>numberOfElements) {
        /*
         * Can opt to choose at max int(numberOfElements) of elements.
         */
        cout << "Number of items can not be greater than total number of elements present in array";
        return;
    }
    sort(arr.begin(), arr.end());
    int j = numberOfElements-1;
    if(N & 1) {
        /*
         * Number of elements to be choose is ODD.
         */
        if(arr[numberOfElements-1] <= 0) {
            /*
             * Start selecting from right margin.
             */
            int count=0;
            for(int i= numberOfElements-1; count < N; i--) {
                pickedNumbers.push_back(arr[i]);
                count++;
            }
            return;
        } else {
            /*
             * choose the last and greatest one and decrease the counter.
             * Now since number of elements to be choosen has become EVEN, move ahead with below logic.
             */
            N -= 1;
            pickedNumbers.push_back(arr[j]);
            j--;
        }
    }
    N = N >> 1; /* Will be choosing the elements in pairs. */
    int i = 0, k = 0, left=0, right=0;
    for(k=0; k<N; k++) {
        left = ( arr[i] * arr[i+1] );
        right = ( arr[j] * arr[j-1] );
        if(left > right) {
            /*
             * choose from left margin
             */
            pickedNumbers.push_back(arr[i]);
            pickedNumbers.push_back(arr[i+1]);
            i += 2;
        } else {
            /*
             * choose from right margin.
             */
            pickedNumbers.push_back(arr[j]);
            pickedNumbers.push_back(arr[j-1]);
            j -= 2;
        }
    }
}

int main() {
    cout << "Please provide the input in the following format.\n\nN (Number of elments in array)\nN integer values of the array elements\n\n";
    int numberOfElements = 0, temp=0;
    cin >> numberOfElements;
    vector<int> arr;
    for (int i=0;i<numberOfElements; i++) {
        cin >> temp;
        arr.push_back(temp);
    }
    int N = 0;
    cin >> N;
    vector<int> pickedNumbers;
    pick_N_numbers(arr, numberOfElements, N, pickedNumbers);
    vector<int>::iterator it;
    for(it=pickedNumbers.begin(); it!=pickedNumbers.end(); it++) {
        cout << (*it) << " ";
    }
    cout << endl;
    return 0;
}

File: Company-Previous-Questions/01/python/n-numbers-with-max-product.py
'''
Name : Jitendra Kumar
Email: jitendra.iitp@gmail.com
'''

def pick_numbers_when_N_is_even(arr, length, N):
    '''
    Handler when N is an even number
    '''
    picked = []
    N = N >> 1 # N = N/2 since I will be choosing the elements in pair of 2
    right = length - 1
    left = 0
    for k in range(N):
        left_product = arr[left] * arr[left+1]
        right_product = arr[right] * arr[right-1]
        if left_product > right_product:
            # Pick 2 numbers from left margin.
            picked.extend([arr[left], arr[left+1]])
            left += 2
        else:
            # Pick 2 numbers from right margin.
            picked.extend([arr[right], arr[right-1]])
            right -= 2
    return picked

def pick_N_numbers(arr, length, N):
    '''
    Pick N numbers such that their product is maximum.
    Form a list of such numbers and return.
    '''
    if N == 0:
        error = "No need to pick any number as expected number of elements to be choosen is 0"
        return error, None

    if N > length:
        error = "Number of items can not be greater than total number of elements present in array"
        return error, None

    arr.sort() # sort the array in ascending order

    if N & 1:
        # When number of elements to be choosen is ODD
        j = length - 1
        if arr[j] <= 0:
            '''
            Start selecting from right margin.
            Since, All of them are <= 0, choosing numbers from right margin
            provides maximum product when they are multiplied.
            '''
            picked = []
            for x in range(N):
                picked.append(arr[j])
                j -= 1
            return None, picked
        else:
            '''
            choose the last and greatest one and decrease the counter.
            Now since number of elements to be choosen has become EVEN, Follow the algorithm for EVEN count.
            '''
            picked = pick_numbers_when_N_is_even(arr, length-1, N-1)
            picked.append(arr[length-1])
            return None, picked
    # When number of elements to be choosen is even
    return None, pick_numbers_when_N_is_even(arr, length, N)

if __name__ == '__main__':
    message = '\nPlease provide the input in the following format.\n\n'\
            'N K (N is number_of_elements, K is number of elements to be choosen)\n'\
            'N space seperated integer values in single line\n'
    print (message)
    try:
        temp1 = list(map(int, input().split()))
        number_of_elements, N = temp1[0], temp1[1]
        if number_of_elements == 0:
            print ('Number of elements in array are 0')
        elif number_of_elements < 0 or N < 0:
            print ('N or K can not be a negative number.')
        else :
            '''
            Just in case number of elements provide as input is greater than number_of_elements
            Choose the first `number_of_elements` numbers.
            '''
            arr = list(map(int, input().split()))[:number_of_elements]
            error, picked = pick_N_numbers(arr, number_of_elements, N)
            if error:
                print (error)
            elif picked:
                print ('The choosen %s numbers are ' % N, end=": ")
                print (*picked)
            else:
                print ('Something went wrong.')
    except ValueError:
        print ('Invalid Input format. Please try again.')
    except Exception as exc:
        print (exc)
